qos.go

```diff
@@ -16,12 +16,188 @@
 
 package goovn
 
+import (
+	"fmt"
+
+	"github.com/ebay/libovsdb"
+)
+
 type QoS struct {
 	UUID       string
 	Priority   int
 	Direction  string
 	Match      string
-	Action     map[string]int
-	Bandwidth  map[string]int
+	Action     map[interface{}]interface{}
+	Bandwidth  map[interface{}]interface{}
 	ExternalID map[interface{}]interface{}
 }
+
+func (odbi *ovnDBImp) rowToQoS(uuid string) *QoS {
+	cacheQoS, ok := odbi.cache[tableQoS][uuid]
+	if !ok {
+		return nil
+	}
+
+	qos := &QoS{
+		UUID:       uuid,
+		Priority:   cacheQoS.Fields["priority"].(int),
+		Direction:  cacheQoS.Fields["direction"].(string),
+		Match:      cacheQoS.Fields["match"].(string),
+		Action:     cacheQoS.Fields["action"].(libovsdb.OvsMap).GoMap,
+		Bandwidth:  cacheQoS.Fields["bandwidth"].(libovsdb.OvsMap).GoMap,
+		ExternalID: cacheQoS.Fields["external_ids"].(libovsdb.OvsMap).GoMap,
+	}
+
+	return qos
+}
+
+func (odbi *ovnDBImp) addQoSImp(ls string, direction string, priority *int, match string, action map[string]int, bandwidth map[string]int, external_ids map[string]string) (*OvnCommand, error) {
+	namedUUID, err := newRowUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	row := make(OVNRow)
+	row["direction"] = direction
+	if priority != nil {
+		row["priority"] = *priority
+	}
+	row["match"] = match
+
+	if action != nil {
+		oMap, err := libovsdb.NewOvsMap(action)
+		if err != nil {
+			return nil, err
+		}
+		row["action"] = oMap
+	}
+	if bandwidth != nil {
+		oMap, err := libovsdb.NewOvsMap(bandwidth)
+		if err != nil {
+			return nil, err
+		}
+		row["bandwidth"] = oMap
+	}
+	if external_ids != nil {
+		oMap, err := libovsdb.NewOvsMap(external_ids)
+		if err != nil {
+			return nil, err
+		}
+		row["external_ids"] = oMap
+	}
+
+	insertOp := libovsdb.Operation{
+		Op:       opInsert,
+		Table:    tableQoS,
+		Row:      row,
+		UUIDName: namedUUID,
+	}
+
+	mutateUUID := []libovsdb.UUID{{namedUUID}}
+	mutateSet, err := libovsdb.NewOvsSet(mutateUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	mutation := libovsdb.NewMutation("qos_rules", opInsert, mutateSet)
+	condition := libovsdb.NewCondition("name", "==", ls)
+
+	mutateOp := libovsdb.Operation{
+		Op:        opMutate,
+		Table:     tableLogicalSwitch,
+		Mutations: []interface{}{mutation},
+		Where:     []interface{}{condition},
+	}
+	operations := []libovsdb.Operation{insertOp, mutateOp}
+
+	return &OvnCommand{operations, odbi, make([][]map[string]interface{}, len(operations))}, nil
+}
+
+func (odbi *ovnDBImp) delQoSImp(ls string, direction string, priority *int, match string) (*OvnCommand, error) {
```

Comment:
why priority is point?  no default value?

Reply:
without pointer we can't allow wildcard removing of qos rules like ovn-nbctl qos-del allows, so we need pointer in this place.

Reply:
OK.